h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,378 @@
+    }
+
+    @Override
+    public void init(boolean expensive) {
+        super.init(expensive);
+        if (_parms.train() != null) {
+            long extensionLevelMax = _parms.train().numCols() - 1;
+            if (_parms.extension_level < 0 || _parms.extension_level > extensionLevelMax) {
+                throw new IllegalStateException("Parameter extension_level must be in interval [0, "
+                        + extensionLevelMax + "] but it is " + _parms.extension_level);
+            }
+            long sampleSizeMax = _parms.train().numRows();
+            if (_parms._sample_size < 0 || _parms._sample_size > sampleSizeMax) {
+                throw new IllegalStateException("Parameter sample_size must be in interval [0, "
+                        + sampleSizeMax + "] but it is " + _parms._sample_size);
+            }
+        }
+    }
+
+    @Override
+    protected double score1(Chunk[] chks, double offset, double weight, double[] fs, int row) {
+        // not used in the current implementation of Extended Isolation Forest
+        return 0;
+    }
+
+    @Override
+    protected Driver trainModelImpl() {
+        return new ExtendedIsolationForestDriver();
+    }
+
+    @Override
+    public ModelCategory[] can_build() {
+        return new ModelCategory[]{
+                ModelCategory.AnomalyDetection
+        };
+    }
+
+    @Override
+    protected ScoreKeeper.ProblemType getProblemType() {
+        return ScoreKeeper.ProblemType.anomaly_detection;
+    }
+
+    @Override
+    public boolean scoreZeroTrees() {
+        return false;
+    }
+
+    @Override
+    public boolean isSupervised() {
+        return false;
+    }
+
+    @Override
+    public boolean havePojo() {
+        return false;
+    }
+
+    @Override
+    public boolean haveMojo() {
+        return true;
+    }
+
+    @Override
+    protected void addCustomInfo(ExtendedIsolationForestModel.ExtendedIsolationForestOutput out) {
+        if (_iTrees != null) {
+            out.iTrees = _iTrees;
+        }
+    }
+
+    private class ExtendedIsolationForestDriver extends Driver {
+
+        @Override
+        protected ExtendedIsolationForestModel makeModel(Key<ExtendedIsolationForestModel> modelKey,
+                                                         ExtendedIsolationForestModel.ExtendedIsolationForestParameters parms) {
+            return new ExtendedIsolationForestModel(modelKey, parms, new ExtendedIsolationForestModel.ExtendedIsolationForestOutput(ExtendedIsolationForest.this));
+        }
+
+        @Override
+        protected boolean doOOBScoring() {
+            return true;
+        }
+
+        @Override
+        protected void scoreAndBuildTrees(boolean oob) {
+            IsolationTreeForkJoinTask [] iTreeTasks = new IsolationTreeForkJoinTask[_parms._ntrees];
+            for (int t = 0; t < _parms._ntrees; t++) {
+                iTreeTasks[t] = new IsolationTreeForkJoinTask();
+                H2O.submitTask(iTreeTasks[t]);
```

Comment:
What is going to happen when we have eg. 10k trees? Do we need to think about this?

Reply:
I don't know how exactly the tasks are distributed around the cluster. No dought that the clever way would be to generate ntrees/numberOfClusters trees in each task. If I am right then the clusters can have different computing power so the logic here can be better than it is now but I don't know how big are the consequences to leaving it as it is.

Reply:
Will be further discussed in https://github.com/h2oai/h2o-3/pull/4907